### ISSUE TITLE
Auto updates on Windows causing service to fail

### ISSUE THREAD

**Issue by robbiemcleod56** [Jackett/Jackett#6948]
The update today caused the Jackett service to fail when starting, producing this error: 

2020-01-14 07:30:43.4170 Error System.IO.FileLoadException: Could not load file or assembly 'YamlDotNet, Version=6.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
File name: 'YamlDotNet, Version=6.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e'
   at Jackett.Common.Services.IndexerManagerService.InitCardigannIndexers(IEnumerable`1 path)
   at Jackett.Common.Services.IndexerManagerService.InitIndexers(IEnumerable`1 path)
   at Jackett.Server.Services.ServerService.Initalize()
   at Jackett.Server.Helper.Initialize()
   at Jackett.Server.Startup.ConfigureServices(IServiceCollection services)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.Initialize()
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()
   at Jackett.Server.Program.Main(String[] args)

The issue was fixed by manually overwriting the files with the contents of the binary zip from releases. Auto update is turned on, and is what caused this. I didn't have to restart the computer to fix the problem. Here is a copy of my update log, everything looks ok. I installed Jackett fresh just a couple days ago, on a system that has next to no software and I would consider it a fresh install, unsure why this is an issue at all. No antivirus software is installed/enabled. 

[updater.txt](https://github.com/Jackett/Jackett/files/4059183/updater.txt)

**Comment by ngosang** [comment]
@junglebus can be related to latest changes?

**Comment by garfield69** [comment]
which particular flavour of windows are you running?
I use win 10 home 1909 and have not had any issues with auto-updates so far.

**Comment by garfield69** [comment]
the updater shows the first error
```
2020-01-14 07:04:22.5832 Error System.IO.IOException: The process cannot access the file 'C:\ProgramData\Jackett\Jackett.Common.dll' because it is being used by another process.
```
clearly some jackett task was still active when the updater was running...

**Comment by garfield69** [comment]
list of locked modules
```
'C:\ProgramData\Jackett\Jackett.Common.dll' 
'C:\ProgramData\Jackett\JackettTray.exe' 
'C:\ProgramData\Jackett\Newtonsoft.Json.dll' 
'C:\ProgramData\Jackett\NLog.dll' 
'C:\ProgramData\Jackett\NLog.Extensions.Logging.dll' 
'C:\ProgramData\Jackett\NLog.Web.AspNetCore.dll' 
'C:\ProgramData\Jackett\System.IO.FileSystem.AccessControl.dll' 
'C:\ProgramData\Jackett\System.Net.Http.dll' 
'C:\ProgramData\Jackett\System.Security.AccessControl.dll' 
'C:\ProgramData\Jackett\System.Security.Principal.Windows.dll' 
'C:\ProgramData\Jackett\System.ServiceProcess.ServiceController.dll' 
```
I wonder if this was caused by user activity of the tray options while the updater was running...an unfortunate case of bad timing? 
perhaps the updater also need to disable the tray menu and enable it afterwards?
just shooting from the hip, no real idea if this is a factor or not.

**Comment by robbiemcleod56** [comment]
I wasn't doing anything with the tray, however it was running when it updated. I kinda wondered why it didn't close. I am running Windows 10 1809 Build 17763.914.

**Comment by ngosang** [comment]
Reinstall Jackett and wait few days until the next automatic update. Maybe it was a one time thing...

**Comment by jparthum** [comment]
Same problem on Windows 7 (ver. 6.1.7601).  

The simplest workaround I found was to restore `C:\ProgramData\Jackett\YamlDotNet.dll` to the previous version:
https://i.imgur.com/OidY7v5.png

**Comment by robbiemcleod56** [comment]
It does seem to be a one time thing. So a few details, maybe it can still be fixed or may be irrelevant. When I installed it I never restarted since it is my file server. After I manually overwrote the files after closing the tray when it stopped working, it worked fine. After a restart, the update seemed to work just fine today, even when the tray was running. Perhaps it's a permissions issue on the first launch? Another thing that changed is that I increased the RAM from 7GB to 14GB. Before that the page file was being hit for almost everything. Other than permissions, I can only imagine lag being the problem. 

Also @jparthum I did overwrite that specific file at first when I experienced the issue, but I had to overwrite all files, not just that one. 

[updater.txt](https://github.com/Jackett/Jackett/files/4089608/updater.txt)

**Comment by garfield69** [comment]
just out of curiosity, have you looked at the system events log (or even the apps events log) for any signs (warning or errors) with problems shutting down the jackettservice.exe, jacketttray.exe or jackettconsole.exe around the time that the updater failed in your first report?

**Comment by robbiemcleod56** [comment]
I found this in event viewer; 

7:03:55AM: Jackett service stopped (for update) 
7:04:53AM: Service started
7:05:26AM: Application: JackettConsole.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.IO.FileLoadException
   at Jackett.Common.Services.IndexerManagerService.InitCardigannIndexers(System.Collections.Generic.IEnumerable`1<System.String>)
   at Jackett.Common.Services.IndexerManagerService.InitIndexers(System.Collections.Generic.IEnumerable`1<System.String>)
   at Jackett.Server.Services.ServerService.Initalize()
   at Jackett.Server.Helper.Initialize()
   at Jackett.Server.Startup.ConfigureServices(Microsoft.Extensions.DependencyInjection.IServiceCollection)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.ConfigureServices(Microsoft.Extensions.DependencyInjection.IServiceCollection)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureApplicationServices()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.Initialize()
   at Microsoft.AspNetCore.Hosting.WebHostBuilder.Build()
   at Jackett.Server.Program.Main(System.String[])
7:05:29AM: Faulting application name: JackettConsole.exe, version: 1.0.0.0, time stamp: 0x99403342
Faulting module name: KERNELBASE.dll, version: 10.0.17763.914, time stamp: 0x6cc9b5fd
Exception code: 0xe0434352
Fault offset: 0x001219b2
Faulting process ID: 0x29f4
Faulting application start time: 0x01d5cad2d2861234
Faulting application path: C:\ProgramData\Jackett\JackettConsole.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report ID: 7050f0b8-ed90-43da-9826-2fc599bc9054
Faulting package full name: 
Faulting package-relative application ID: 
7:09:39AM: Fault bucket 2126082669088568796, type 5
Event Name: CLR20r3
Response: Not available
Cab Id: 0

Problem signature:
P1: JackettConsole.exe
P2: 1.0.0.0
P3: 99403342
P4: Jackett.Common
P5: 0.12.1539.0
P6: bb019ce5
P7: e1
P8: 6
P9: System.IO.FileLoadException
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2104.tmp.dmp
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2C7F.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2EF1.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2EEF.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER2F8C.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_JackettConsole.e_386ab773736a3eefda58158a014dcdfc55bd336_9ad583c6_0ae04381

Analysis symbol: 
Rechecking for solution: 0
Report Id: 7050f0b8-ed90-43da-9826-2fc599bc9054
Report Status: 268435456
Hashed bucket: 2c4a90422b404173fd815cec4e2375dc
Cab Guid: 0
7:05:40: Service stopped

**Comment by garfield69** [comment]
which confirms your report, in that the service was shutdown at 7:03:55 and restarted at 7:04:53,
but no other events occurred in between those times that the system felt were significant to report, that could have provided an explanation for why the updater failed to complete the update, other than what I found in the updater.txt log (the file locks on the  jacketttray module and open libraries).
Ah well, nice try but no banana ;-)

**Comment by jparthum** [comment]
@robbiemcleod56
Where did you get the new `YamlDotNet.dll` file you used to overwrite the existing one?
I'm not sure what this is referring to:
> ... binary zip from releases.

**Comment by robbiemcleod56** [comment]
@jparthum 
The Jackett releases page.  I grabbed the appropriate version matching the update and first tried just that file, but had to close the tray and overwrite all the contents of the zip to C:\ProgramData\Jackett.  I grabbed Jackett.Binaries.Windows.zip from here. 
https://github.com/Jackett/Jackett/releases/tag/v0.12.1572

**Comment by jparthum** [comment]
Ah OK, I didn't realize files from outside of this repository were being distributed in `Jackett.Binaries.Windows.zip`.

If this helps provide any insight; v0.12.1572 seems to have been distributed with `YamlDotNet.dll` ver. **8.0.0.0**, but `Jackett.Common.csproj` still referenced `YamlDotNet.dll` ver. **6.0.0.0** [until v0.12.1623](https://github.com/Jackett/Jackett/commit/dc87b36ae0e6f407ee0129e5d373ef50403efb5d#diff-de9a1ebf59afeb7d7d472061eb4b18a8).

**Comment by junglebus** [comment]
@jparthum 
There is no incompatibility between Jackett and YamlDotNet 8 (Tens of thousands of downloads of the Windows version https://somsubhra.com/github-release-stats/?username=jackett&repository=Jackett)
The library did get updated recently so suspect a file is locked/antivirus and not getting updated
- Stop the Jackett Service
-Make sure the Jackett Tray is not running
-Check task manager to make sure no instances of Jackett processes are running
-Then run the latest Jackett Installer and it will work

**Comment by jparthum** [comment]
@junglebus 
I didn't use the term 'incompatibility' so I'm not quite sure what to make of your response.

Is the information I posted incorrect?

**Comment by junglebus** [comment]
@jparthum I assumed that you are implying there is an incompatibility from the below statement.
And, yes, what you have posted is incorrect

> If this helps provide any insight; v0.12.1572 seems to have been distributed with `YamlDotNet.dll` ver. **8.0.0.0**, but `Jackett.Common.csproj` still referenced `YamlDotNet.dll` ver. **6.0.0.0** [until v0.12.1623](https://github.com/Jackett/Jackett/commit/dc87b36ae0e6f407ee0129e5d373ef50403efb5d#diff-de9a1ebf59afeb7d7d472061eb4b18a8).

If you download v0.12.1572 https://github.com/Jackett/Jackett/releases/download/v0.12.1572/Jackett.Binaries.Windows.zip
The DLL is 8.0.0 and Jackett.Common references 8.0.0

![image](https://user-images.githubusercontent.com/22320807/73163599-708b9980-4144-11ea-8e23-0162324abda7.png)

Open `JackettConsole.deps.json`
![image](https://user-images.githubusercontent.com/22320807/73163548-5487f800-4144-11ea-9bc2-4d15429239ad.png)

Also, if you look at the source code at the time of v0.12.1572 https://github.com/Jackett/Jackett/blob/v0.12.1572/src/Jackett.Common/Jackett.Common.csproj
![image](https://user-images.githubusercontent.com/22320807/73163714-a29cfb80-4144-11ea-8369-e0472f78d102.png)

If you try v0.12.1561, you'll see the same thing all 8.0.0's

**Comment by jparthum** [comment]
My apologies @junglebus ; I apparently lost track of versioning while downloading (renaming incorrectly) and confused v0.12.1561 with v0.12.1539.
My intention was merely to help diagnose and I thought I had stumbled across something potentially useful.  Sorry for the confusion.

**Comment by robbiemcleod56** [comment]
@junglebus 
Yes, the tray wasn't closed automatically so the files related to that part of it were locked. I had installed the program originally with the exe installer, and it went to update a couple days later but it didn't close the tray. That's what I'm trying to figure out, is why that part of it didn't close when the service had. It's worked just fine since I manually closed it and updated from the zip file, all updates are fine right now, I'm running 0.12.1638.0 currently. This ONLY happened after the initial install.
